sdk/lib/cmlib/cmcheck.c

```diff
@@ -0,0 +1,1800 @@
+        }
+
+        /* Check that the value size is sane */
+        ValueDataSize = HvGetCellSize(Hive, ValueData);
+        ValueNameLength = ValueData->u.KeyValue.NameLength;
+        TotalValueNameLength = ValueNameLength + FIELD_OFFSET(CM_KEY_VALUE, Name);
+        if (TotalValueNameLength > ValueDataSize)
+        {
+            /* Kill this value from the list */
+            if (!CmpRepairValueListCount(Hive,
+                                         CurrentCell,
+                                         ListCountIndex,
+                                         ValueListData))
+            {
+                DPRINT1("CmpValidateValueListByCount(): The total size is bigger than the actual cell size (total size %lu, cell size %lu, at index %lu)\n",
+                        TotalValueNameLength, ValueDataSize, ListCountIndex);
+                return CM_CHECK_REGISTRY_VALUE_CELL_SIZE_NOT_SANE;
+            }
+
+            /*
+             * We have altered the list count of the cell,
+             * now we must update the list count here as well.
+             * Decrease the count and continue with the next value.
+             * Inform the caller we have removed a value.
+             */
+            ListCount--;
+            *ValuesRemoved++;
+            DPRINT1("CmpValidateValueListByCount(): Damaged value removed, continuing with the next value...\n");
+            continue;
+        }
+
+        /*
+         * The value cell has a sane size. The last thing
+         * to validate is the actual data of the value cell.
+         * That is, we want that the data itself and length
+         * are consistent. Technically speaking, value keys
+         * that are small are directly located in the value
+         * cell and it's built-in, in other words, the data
+         * is immediately present in the cell so we don't have
+         * to bother validating them since they're alright on
+         * their own. This can't be said the same about normal
+         * values though.
+         */
+        DataCell = ValueData->u.KeyValue.Data;
+        if (!CmpIsKeyValueSmall(&DataSize, ValueData->u.KeyValue.DataLength))
+        {
+            /*
+             * Validate the actual data based on size.
+             *
+             * CASE DataSize == 0 && DataCell != HCELL_NIL:
+             *
+             * - The length is 0 but there's some data still. This
+             * could indicate this registry hive had some data before
+             * but its internal properties got corrupt and the data
+             * is still here. We cannot trust it so ultimately the
+             * value will be killed.
+             *
+             * CASE DataSize > 0:
+             *
+             * - This registry hive still has some data as the length
+             * tells us but we want to be sure this kind of data is
+             * not "cancerous", aka, is the cell allocated?
+             *
+             */
```

Comment:
One superfluous line even in your commenting style. At least 3 too much for my style.